File: DS-Algos/DSU.cpp
#include <bits/stdc++.h>
using namespace std;
class DSU
{
public:
    vector<int> parent;
    vector<int> rank;
    DSU(int n)
    {
        parent = vector<int>(n, -1);
        rank = vector<int>(n, 1);
    }
    int find(int a)
    {
        if (parent[a] == -1)
            return a;
        return parent[a] = find(parent[a]);
    }
    bool unionbyrank(int a, int b)
    {
        int pa = find(a);
        int pb = find(b);
        if (pa == pb)
            return true;
        if (rank[pa] > rank[pb])
        {
            parent[pb] = pa;
            rank[pa] += rank[pb];
        }
        else
        {
            parent[pa] = pb;
            rank[pb] += rank[pa];
        }
        return true;
    }
};

File: DS-Algos/DSU.py
class DSU:
    def __init__(self, n):
        self.parent = [-1] * n
        self.rank = [1] * n

    def union(self, a, b):
        pa, pb = self.find(a), self.find(b)
        if pa == pb:
            return False
        if self.rank[pa] > self.rank[pb]:
            self.parent[pb] = pa
            self.rank[pa] += self.rank[pb]
        else:
            self.parent[pa] = pb
            self.rank[pb] += self.rank[pa]
        return True

    def find(self, a):
        if self.parent[a] == -1:
            return a
        pa = self.find(self.parent[a])
        self.parent[a] = pa
        return pa
